chain-extensions/pallet-assets/src/lib.rs

```diff
@@ -18,398 +18,225 @@
-
-    fn try_from(value: u16) -> Result<Self, Self::Error> {
-        match value {
-            1 => Ok(AssetsFunc::Create),
-            2 => Ok(AssetsFunc::Transfer),
-            3 => Ok(AssetsFunc::Mint),
-            4 => Ok(AssetsFunc::Burn),
-            5 => Ok(AssetsFunc::BalanceOf),
-            6 => Ok(AssetsFunc::TotalSupply),
-            7 => Ok(AssetsFunc::Allowance),
-            8 => Ok(AssetsFunc::ApproveTransfer),
-            9 => Ok(AssetsFunc::CancelApproval),
-            10 => Ok(AssetsFunc::TransferApproved),
-            11 => Ok(AssetsFunc::SetMetadata),
-            12 => Ok(AssetsFunc::MetadataName),
-            13 => Ok(AssetsFunc::MetadataSymbol),
-            14 => Ok(AssetsFunc::MetadataDecimals),
-            15 => Ok(AssetsFunc::TransferOwnership),
-            _ => Err(DispatchError::Other(
-                "PalletAssetsExtension: Unimplemented func_id",
-            )),
-        }
-    }
-}
+type Weight<T> = <T as pallet_assets::Config>::WeightInfo;
 
 /// Pallet Assets chain extension.
-pub struct AssetsExtension<T, W>(PhantomData<(T, W)>);
+pub struct AssetsExtension<T>(PhantomData<T>);
 
-impl<T, W> Default for AssetsExtension<T, W> {
+impl<T> Default for AssetsExtension<T> {
     fn default() -> Self {
         AssetsExtension(PhantomData)
     }
 }
 
-impl<T, W> ChainExtension<T> for AssetsExtension<T, W>
+impl<T> ChainExtension<T> for AssetsExtension<T>
 where
     T: pallet_assets::Config + pallet_contracts::Config,
     <T as pallet_assets::Config>::AssetId: Copy,
     <<T as SysConfig>::Lookup as StaticLookup>::Source: From<<T as SysConfig>::AccountId>,
-    <T as SysConfig>::AccountId: From<[u8; 32]>,
-    W: weights::WeightInfo,
 {
     fn call<E: Ext>(&mut self, env: Environment<E, InitState>) -> Result<RetVal, DispatchError>
     where
         E: Ext<T = T>,
     {
-        let func_id = env.func_id().try_into()?;
         let mut env = env.buf_in_buf_out();
-
-        match func_id {
-            AssetsFunc::Create => {
-                let (origin, id, admin, min_balance): (
-                    Origin,
+        match env.func_id().try_into().map_err(|_| {
+            DispatchError::Other("Unsupported func id in Pallet Assets Chain Extension")
+        })? {
+            Transfer => {
+                let (id, target, amount): (
                     <T as pallet_assets::Config>::AssetId,
                     T::AccountId,
                     T::Balance,
                 ) = env.read_as()?;
 
-                let base_weight = <T as pallet_assets::Config>::WeightInfo::create();
-                env.charge_weight(base_weight)?;
+                log::trace!(target: "pallet-chain-extension-assets::transfer",
+                    "Raw arguments: id: {:?}, to: {:?}, amount: {:?}",
+                      id, target, amount);
 
-                let raw_origin = select_origin!(&origin, env.ext().address().clone());
+                env.charge_weight(Weight::<T>::transfer())?;
```

Comment:
General comment - shouldn't the `charge_weight` call be executed as the first thing?